# Exercise 4.8: TASK

# 01 Import Libraries

In [3]:
# Importing Pandas, NumPy and OS
import pandas as pd
import numpy as np
import os

# 02 Import Dataframe

## Path creation

In [6]:
path = r'/Users/Andreas/Desktop/13-02-2025 Instacart Basket Analysis/02 Data'

In [7]:
#checking output
path

'/Users/Andreas/Desktop/13-02-2025 Instacart Basket Analysis/02 Data'

## Import of Dataframe

In [9]:
#importing dataframe orders_products_merged2.pkl
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged2.pkl'))

In [10]:
# checking shape
ords_prods_merge.shape

(32404859, 18)

In [11]:
#checking output
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id    prices  \
0           1  Chocolate Sandwich Cookies        61             19  5.800781   
1           1  Chocolate Sandwich Cookies        61             19  5.800781   
2           1  Chocolate Sandwich Cookies        61             19  5.800781   
3           1  Chocolate Sandwich Cookies        61             19  5.800781   
4           1  Chocolate Sandwich Cookies        61             19  5.800781   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_without_ordering  add_to_cart_order  reordered _merge  \
0                    3.0                  5          0   both   
1                   20.0                  1          1   both   
2                    6.0                 20          0   both   
3                    0.0                 10          0   both   
4                    9.0                 11          1   both   

         price_label     busiest_day         busiest_days  \
0  Mid-range product  Regularly busy  Regularly busy days   
1  Mid-range product  Regularly busy  Regularly busy days   
2  Mid-range product     Busiest day         Busiest days   
3  Mid-range product  Regularly busy      Least busy days   
4  Mid-range product      Least busy      Least busy days   

  busiest_period_of_day  
0           Most Orders  
1        Average Orders  
2        Average Orders  
3           Most Orders  
4        Average Orders

# Question 2

    In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.


In [14]:
#grouping all ords_prods_merge dataframe by department id and agg by mean of order number
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

# Question 3

Analyze the result. How do the results for the entire dataframe differ from those of the subset? 

## Comparison of SUBSET from Exercise and ords_prods_merge
    SUBSET                                             
    department_id
    1     15.577493
    2     17.320781
    3     16.084944
    4     17.530458
    5     14.763075
    6     16.658449
    7     17.031590
    8     15.076662
    9     15.447580
    10    18.681852
    11    15.447411
    12    14.327957
    13    16.548642
    14    16.960241
    15    16.121948
    16    17.803851
    17    15.593633
    18    19.674252
    19    16.899756
    20    16.255442
    21    25.535479
    Name: order_number, dtype: float64

    ords_rods_merge
    department_id
    1     15.457838
    2     17.277920
    3     17.170395
    4     17.811403
    5     15.215751
    6     16.439806
    7     17.225802
    8     15.340650
    9     15.895474
    10    20.197148
    11    16.170638
    12    15.887671
    13    16.583536
    14    16.773669
    15    16.165037
    16    17.665606
    17    15.694469
    18    19.310397
    19    17.177343
    20    16.473447
    21    22.902379
    Name: order_number, dtype: float64

## Answer

They are very similar none are identical but the subset is fairly representative of the ords_prods_merge Dataframe this is understandable as it is a sample of the data which pulls 1 million rows from 32 million of the total Dataframe of ords_prods_merge. 

# Question 4

Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

## Transform

In [21]:
#1.Split the data into groups based on the “user_id” column.

#2.Apply the transform() function on the “order_number” column to generate the maximum orders for each user.

#3. Create a new column, “max_order,” into which it will place the results of your aggregation.

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/bz/xnfhbvns50zc0x2wgk6fnbg80000gq/T/ipykernel_70580/4177515964.py:7: FutureWarning: The provided callable <function max at 0x11158e0c0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [22]:
# check output
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id    prices  \
0           1  Chocolate Sandwich Cookies        61             19  5.800781   
1           1  Chocolate Sandwich Cookies        61             19  5.800781   
2           1  Chocolate Sandwich Cookies        61             19  5.800781   
3           1  Chocolate Sandwich Cookies        61             19  5.800781   
4           1  Chocolate Sandwich Cookies        61             19  5.800781   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_without_ordering  add_to_cart_order  reordered _merge  \
0                    3.0                  5          0   both   
1                   20.0                  1          1   both   
2                    6.0                 20          0   both   
3                    0.0                 10          0   both   
4                    9.0                 11          1   both   

         price_label     busiest_day         busiest_days  \
0  Mid-range product  Regularly busy  Regularly busy days   
1  Mid-range product  Regularly busy  Regularly busy days   
2  Mid-range product     Busiest day         Busiest days   
3  Mid-range product  Regularly busy      Least busy days   
4  Mid-range product      Least busy      Least busy days   

  busiest_period_of_day  max_order  
0           Most Orders         32  
1        Average Orders         32  
2        Average Orders          5  
3           Most Orders          3  
4        Average Orders          3

## Creating Loyalty Flag for Transformed data

In [24]:
# locate max order in ords_prods_merge if its greater than 40 in column loyalty_flag return loyal customer
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal Customer'

# locate max order in ords_prods_merge if its <= 40  and > 10 in column loyalty_flag return regular customer
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order']> 10), 'loyalty_flag'] = 'Regular Customer'

# locate max order in ords_prods_merge if its < 10 in column loyalty_flag return new customer
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New Customer'

In [25]:
# check output
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id    prices  \
0           1  Chocolate Sandwich Cookies        61             19  5.800781   
1           1  Chocolate Sandwich Cookies        61             19  5.800781   
2           1  Chocolate Sandwich Cookies        61             19  5.800781   
3           1  Chocolate Sandwich Cookies        61             19  5.800781   
4           1  Chocolate Sandwich Cookies        61             19  5.800781   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_without_ordering  add_to_cart_order  reordered _merge  \
0                    3.0                  5          0   both   
1                   20.0                  1          1   both   
2                    6.0                 20          0   both   
3                    0.0                 10          0   both   
4                    9.0                 11          1   both   

         price_label     busiest_day         busiest_days  \
0  Mid-range product  Regularly busy  Regularly busy days   
1  Mid-range product  Regularly busy  Regularly busy days   
2  Mid-range product     Busiest day         Busiest days   
3  Mid-range product  Regularly busy      Least busy days   
4  Mid-range product      Least busy      Least busy days   

  busiest_period_of_day  max_order      loyalty_flag  
0           Most Orders         32  Regular Customer  
1        Average Orders         32  Regular Customer  
2        Average Orders          5      New Customer  
3           Most Orders          3      New Customer  
4        Average Orders          3      New Customer

In [26]:
# checking frequency
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular Customer    15876776
Loyal Customer      10284093
New Customer         6243990
Name: count, dtype: int64

# Question 5

    The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

## PROBLEM
I found inf values in my prices column only which accounted for 698 rows to address this problem I would contact client to send the relevant information, however, since this is an exercise i chose to change the inf values into NaN as to not affect the data and since its only 698 rows from 32+million rows it won't affect the data integrity

In [30]:
# ords_prods_merge INF values changed to NaN
ords_prods_merge.replace([np.inf, -np.inf], np.nan, inplace=True)

In [31]:
# descriptive statistics for prices by loyalty category 
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal Customer     9.578717  1.0  14896.0
New Customer      10.059199  1.0  14896.0
Regular Customer   9.894029  1.0  14896.0

# Question 6

    The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
    If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.” If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [34]:
#1.Split the data into groups based on the “user_id” column.

#2.Apply the transform() function on the “prices” column to generate the average price per item for each user.

#3. Create a new column, “user_mean_price,” into which it will place the results of the aggregation.
ords_prods_merge['user_mean_price'] = ords_prods_merge.groupby('user_id')['prices'].transform('mean')

In [35]:
#check output
ords_prods_merge['user_mean_price']

0           6.936029
1           6.936029
2           7.930211
3           4.972421
4           4.972421
              ...   
32404854    8.220413
32404855    7.364667
32404856    8.499795
32404857    8.499795
32404858    8.133077
Name: user_mean_price, Length: 32404859, dtype: float64

In [36]:
# locate user mean price in ords_prods_merge if its < 10 in column Spender habit return Low Spender
ords_prods_merge.loc[ords_prods_merge['user_mean_price'] < 10, 'Spender_habit'] = 'Low Spender'

# locate user mean price in ords_prods_merge if its >= 10 in column Sender habit return High Spender
ords_prods_merge.loc[ords_prods_merge['user_mean_price'] >= 10, 'Spender_habit'] = 'High Spender'

In [37]:
#output check
ords_prods_merge['Spender_habit'].value_counts(dropna = False)

Spender_habit
Low Spender     31828319
High Spender      576540
Name: count, dtype: int64

In [38]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id    prices  \
0           1  Chocolate Sandwich Cookies        61             19  5.800781   
1           1  Chocolate Sandwich Cookies        61             19  5.800781   
2           1  Chocolate Sandwich Cookies        61             19  5.800781   
3           1  Chocolate Sandwich Cookies        61             19  5.800781   
4           1  Chocolate Sandwich Cookies        61             19  5.800781   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  reordered  _merge        price_label     busiest_day  \
0  ...          0    both  Mid-range product  Regularly busy   
1  ...          1    both  Mid-range product  Regularly busy   
2  ...          0    both  Mid-range product     Busiest day   
3  ...          0    both  Mid-range product  Regularly busy   
4  ...          1    both  Mid-range product      Least busy   

          busiest_days busiest_period_of_day max_order      loyalty_flag  \
0  Regularly busy days           Most Orders        32  Regular Customer   
1  Regularly busy days        Average Orders        32  Regular Customer   
2         Busiest days        Average Orders         5      New Customer   
3      Least busy days           Most Orders         3      New Customer   
4      Least busy days        Average Orders         3      New Customer   

   user_mean_price Spender_habit  
0         6.936029   Low Spender  
1         6.936029   Low Spender  
2         7.930211   Low Spender  
3         4.972421   Low Spender  
4         4.972421   Low Spender  

[5 rows x 22 columns]

# Question 7

    In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
    
    If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
    
    If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
    
    If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”


In [42]:
#1.Split the data into groups based on the “user_id” column.

#2.Apply the transform() function on the “days_without_ordering” column to generate the median days without ordering for each user.

#3. Create a new column, “median_days_no_order,” into which it will place the results of the aggregation.
ords_prods_merge['median_days_no_order'] = ords_prods_merge.groupby('user_id')['days_without_ordering'].transform('median')

In [44]:
#checking output
ords_prods_merge['median_days_no_order']

0            8.0
1            8.0
2            6.0
3            9.0
4            9.0
            ... 
32404854     7.0
32404855    30.0
32404856     6.0
32404857     6.0
32404858     6.0
Name: median_days_no_order, Length: 32404859, dtype: float16

In [48]:
# locate median days no order in ords_prods_merge if its > 20 in column Customer Type return Non-Frequent Customer
ords_prods_merge.loc[ords_prods_merge['median_days_no_order'] > 20, 'Customer Type'] = 'Non-Frequent Customer'
# locate median days no order in ords_prods_merge if its <= 20 & >10 in column Customer Type return Regular Customer
ords_prods_merge.loc[(ords_prods_merge['median_days_no_order'] <= 20) & (ords_prods_merge['median_days_no_order'] > 10), 'Customer Type'] = 'Regular Customer'
# locate median days no order in ords_prods_merge if its <= 10 in column Customer Type return Frequent Customer
ords_prods_merge.loc[ords_prods_merge['median_days_no_order'] <= 10, 'Customer Type'] = 'Frequent Customer'

In [50]:
ords_prods_merge['Customer Type'].value_counts(dropna = False)

Customer Type
Frequent Customer        22796659
Regular Customer          6921472
Non-Frequent Customer     2686728
Name: count, dtype: int64

In [54]:
#check output
ords_prods_merge.head(10)

product_id                product_name  aisle_id  department_id    prices  \
0           1  Chocolate Sandwich Cookies        61             19  5.800781   
1           1  Chocolate Sandwich Cookies        61             19  5.800781   
2           1  Chocolate Sandwich Cookies        61             19  5.800781   
3           1  Chocolate Sandwich Cookies        61             19  5.800781   
4           1  Chocolate Sandwich Cookies        61             19  5.800781   
5           1  Chocolate Sandwich Cookies        61             19  5.800781   
6           1  Chocolate Sandwich Cookies        61             19  5.800781   
7           1  Chocolate Sandwich Cookies        61             19  5.800781   
8           1  Chocolate Sandwich Cookies        61             19  5.800781   
9           1  Chocolate Sandwich Cookies        61             19  5.800781   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   
5   1701441      777            16                   1                  7   
6   1871483      825             3                   2                 14   
7   1290456      910            12                   3                 10   
8    369558     1052            10                   1                 20   
9    589712     1052            15                   1                 12   

   ...        price_label     busiest_day         busiest_days  \
0  ...  Mid-range product  Regularly busy  Regularly busy days   
1  ...  Mid-range product  Regularly busy  Regularly busy days   
2  ...  Mid-range product     Busiest day         Busiest days   
3  ...  Mid-range product  Regularly busy      Least busy days   
4  ...  Mid-range product      Least busy      Least busy days   
5  ...  Mid-range product  Regularly busy         Busiest days   
6  ...  Mid-range product  Regularly busy  Regularly busy days   
7  ...  Mid-range product  Regularly busy      Least busy days   
8  ...  Mid-range product  Regularly busy         Busiest days   
9  ...  Mid-range product  Regularly busy         Busiest days   

  busiest_period_of_day max_order      loyalty_flag user_mean_price  \
0           Most Orders        32  Regular Customer        6.936029   
1        Average Orders        32  Regular Customer        6.936029   
2        Average Orders         5      New Customer        7.930211   
3           Most Orders         3      New Customer        4.972421   
4        Average Orders         3      New Customer        4.972421   
5        Average Orders        26  Regular Customer        6.934674   
6           Most Orders         9      New Customer        5.957090   
7           Most Orders        12  Regular Customer        6.679492   
8        Average Orders        20  Regular Customer        7.162231   
9           Most Orders        20  Regular Customer        7.162231   

  Spender_habit  median_days_no_order      Customer Type  
0   Low Spender                   8.0  Frequent Customer  
1   Low Spender                   8.0  Frequent Customer  
2   Low Spender                   6.0  Frequent Customer  
3   Low Spender                   9.0  Frequent Customer  
4   Low Spender                   9.0  Frequent Customer  
5   Low Spender                  11.0   Regular Customer  
6   Low Spender                  19.0   Regular Customer  
7   Low Spender                   3.5  Frequent Customer  
8   Low Spender                   9.5  Frequent Customer  
9   Low Spender                   9.5  Frequent Customer  

[10 rows x 24 columns]

In [58]:
# drop column busiest day unnecessary 
ords_prods_merge.drop(columns = 'busiest_day')

product_id                product_name  aisle_id  department_id  \
0                  1  Chocolate Sandwich Cookies        61             19   
1                  1  Chocolate Sandwich Cookies        61             19   
2                  1  Chocolate Sandwich Cookies        61             19   
3                  1  Chocolate Sandwich Cookies        61             19   
4                  1  Chocolate Sandwich Cookies        61             19   
...              ...                         ...       ...            ...   
32404854       49688      Fresh Foaming Cleanser        73             11   
32404855       49688      Fresh Foaming Cleanser        73             11   
32404856       49688      Fresh Foaming Cleanser        73             11   
32404857       49688      Fresh Foaming Cleanser        73             11   
32404858       49688      Fresh Foaming Cleanser        73             11   

             prices  order_id  user_id  order_number  orders_day_of_week  \
0          5.800781   3139998      138            28                   6   
1          5.800781   1977647      138            30                   6   
2          5.800781    389851      709             2                   0   
3          5.800781    652770      764             1                   3   
4          5.800781   1813452      764             3                   4   
...             ...       ...      ...           ...                 ...   
32404854  13.500000   1788356   200215             2                   0   
32404855  13.500000   3401313   200377             1                   4   
32404856  13.500000    809510   200873             5                   3   
32404857  13.500000   2359893   200873             9                   3   
32404858  13.500000   2385091   205926            11                   1   

          order_hour_of_day  ...  _merge        price_label  \
0                        11  ...    both  Mid-range product   
1                        17  ...    both  Mid-range product   
2                        21  ...    both  Mid-range product   
3                        13  ...    both  Mid-range product   
4                        17  ...    both  Mid-range product   
...                     ...  ...     ...                ...   
32404854                  9  ...    both  Mid-range product   
32404855                 11  ...    both  Mid-range product   
32404856                  8  ...    both  Mid-range product   
32404857                 15  ...    both  Mid-range product   
32404858                 15  ...    both  Mid-range product   

                 busiest_days busiest_period_of_day max_order  \
0         Regularly busy days           Most Orders        32   
1         Regularly busy days        Average Orders        32   
2                Busiest days        Average Orders         5   
3             Least busy days           Most Orders         3   
4             Least busy days        Average Orders         3   
...                       ...                   ...       ...   
32404854         Busiest days           Most Orders         6   
32404855      Least busy days           Most Orders         4   
32404856      Least busy days        Average Orders        20   
32404857      Least busy days           Most Orders        20   
32404858         Busiest days           Most Orders        11   

              loyalty_flag user_mean_price  Spender_habit  \
0         Regular Customer        6.936029    Low Spender   
1         Regular Customer        6.936029    Low Spender   
2             New Customer        7.930211    Low Spender   
3             New Customer        4.972421    Low Spender   
4             New Customer        4.972421    Low Spender   
...                    ...             ...            ...   
32404854      New Customer        8.220413    Low Spender   
32404855      New Customer        7.364667    Low Spender   
32404856  Regular Customer        8.499795    Low Spender   
32404857  Regular Custom

In [62]:
ords_prods_merge.shape

(32404859, 24)

# Export

In [66]:
ords_prods_merge.to_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merge3.pkl'))